### PR TITLE
 Tidy up of crate/module API: Stage 7

### PR DESCRIPTION
**Pull request by horned-sphere**
Cleans up the structure of `swimos_agent`. Also attemptes to reduce the exported surfaces.

Adds basic doc comments.

Lots more could be done with this module but it would take too long.

### CODE REVIEW

File: server/swimos_agent/src/agent_lifecycle/utility/downlink_builder/map.rs [hunk 1]
```diff
@@ -81,7 +77,10 @@ impl<Context, K, V, LC> StatelessMapDownlinkBuilder<Context, K, V, LC>
 where
     LC: StatelessMapLifecycle<Context, K, V>,
 {
-    /// Specify a new event handler to be executed when the downlink enters the linked state.
+    /// Specify a handler for the `on_linked` event.
+    ///
+    /// # Arguments
+    /// * `handler` - The event handler.
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_lifecycle/utility/downlink_builder/map.rs:83]: None of the functions below have an argument of 'handler', they are all 'f'

File: server/swimos_agent/src/agent_lifecycle/utility/downlink_builder/map.rs [hunk 2]
```diff
@@ -331,7 +354,10 @@ where
         }
     }
 
-    /// Specify a new event handler to be executed when the downlink enters the unlinked state.
+    /// Specify a handler for the `on_linked` event.
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_lifecycle/utility/downlink_builder/map.rs:357]:
```suggestion
    /// Specify a handler for the `on_unlinked` event.
```

File: server/swimos_agent/src/agent_lifecycle/utility/mod.rs [hunk 3]
```diff
@@ -421,7 +439,17 @@ impl<Agent: 'static> HandlerContext<Agent> {
         run_schedule(schedule)
     }
 
-    /// Run a (potentially infinite) sequence of futures (each resulting in an event handler).
+    /// Schedule a (potentially infinite) sequence of [`EventHandler`]s to run on a schedule. For each pair of a
+    /// delay and and [`EventHandler`] returned by the provided async [`Stream`], the handler is scheduled to run
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_lifecycle/utility/mod.rs:443]: This documentation doesn't seem right. This returns a handler which polls a stream which yields event handlers?

File: server/swimos_agent/src/agent_lifecycle/utility/mod.rs [hunk 4]
```diff
@@ -431,6 +459,15 @@ impl<Agent: 'static> HandlerContext<Agent> {
...
     /// Schedule a (potentially infinite) sequence of handlers to run with a fixed delay between them.
+    ///
+    /// # Note
+    ///
+    /// Both the iterator and the handlers must be [`Send`] as the task running the agent could be moved
+    /// to another thread while they are still in use.
+    ///
+    /// # Arguments
+    /// * `delay` - The fixed delay.
+    /// * `schedule` - An iterator returning a sequence of pairs of delays and handlers.
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_lifecycle/utility/mod.rs:470]:
```suggestion
    /// * `handlers` - An iterator returning a sequence of pairs of delays and handlers.
```

File: server/swimos_agent/src/agent_lifecycle/utility/mod.rs [hunk 5]
```diff
@@ -444,9 +481,18 @@ impl<Agent: 'static> HandlerContext<Agent> {
...
     }
 
-    /// Schedule a (potentially infinite) sequence of futures (each resulting in an event handler) to
+    /// Schedule a (potentially infinite) sequence of futures (each resulting in an [`EventHandler`]) to
     /// run with a fixed delay between them. The delay is computed when the future starts executing so if
     /// a futures takes longer than the delay to complete, the next will start immediately.
+    ///
+    /// # Note
+    ///
+    /// All of the iterator, the futures and the handlers must be [`Send`] as the task running the agent
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_lifecycle/utility/mod.rs:490]:
```suggestion
    /// The iterator, futures and handlers must be [`Send`] as the task running the agent
```

File: server/swimos_agent/src/agent_lifecycle/utility/mod.rs [hunk 6]
```diff
@@ -503,6 +568,10 @@ impl<Agent: 'static> HandlerContext<Agent> {
     }
 
     /// Suspend a future to be executed by the agent task.
+    /// # Note
+    ///
+    /// The future must be[`Send`] as the task running the agent could be moved to another thread while it
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_lifecycle/utility/mod.rs:573]:
```suggestion
    /// The future must be [`Send`] as the task running the agent could be moved to another thread while it
```

File: server/swimos_agent/src/agent_model/downlink/mod.rs [hunk 7]
```diff
@@ -272,3 +252,73 @@ where
         }
     }
 }
+
+/// Indication that the downlink task has completed some unit of work.
+#[derive(Debug, PartialEq, Eq)]
+pub enum DownlinkChannelEvent {
+    /// An change has occurred to the state of the downlink (either an event or a change to
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_model/downlink/mod.rs:259]:
```suggestion
    /// A change has occurred to the state of the downlink (either an event or a change to
```

File: server/swimos_agent/src/agent_model/downlink/mod.rs [hunk 8]
```diff
@@ -272,3 +252,73 @@ where
...
+    #[error("Reading from the downlink input failed.")]
+    ReadFailed,
+    /// Attempting to write to the downlink failed.
+    #[error("Writing to the downlink output failed: {0}")]
+    WriteFailed(#[from] std::io::Error),
+}
+
+/// Allows a downlink to be driven by an agent task, without any knowledge of the types used
+/// internally by the downlink. A [`DownlinkChannel`] is essentially a stream of event handlers.
+/// However, the operation to get the next handler is split across two methods(on to wait for
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_model/downlink/mod.rs:285]:
```suggestion
/// However, the operation to get the next handler is split across two methods (one to wait for
```

File: server/swimos_agent/src/agent_model/downlink/mod.rs [hunk 9]
```diff
@@ -272,3 +252,73 @@ where
...
+/// an incoming message and the other to create an event handler, where appropriate). This split
+/// is necessary to avoid holding a reference to the downlink lifecycle across an await point.
+/// Otherwise, we would need to add a bound that the lifecycles must be [`Sync`] which
+/// is undesirable as it would prevent the use of [`std::cell::RefCell`] in lifecycle fields.
+///
+/// #Type Arguments
+///
+/// * `Context` - The context within which the event handlers must be run (typically an agent type).
+pub trait DownlinkChannel<Context> {
+    /// Attach the downlink to it's input and output streams. This can be called multiple times
```
Comment (SirCipher) [inline server/swimos_agent/src/agent_model/downlink/mod.rs:295]:
```suggestion
    /// Attach the downlink to its input and output streams. This can be called multiple times
```

File: server/swimos_agent/src/event_handler/mod.rs [hunk 10]
```diff
@@ -103,6 +104,9 @@ where
     }
 }
 
+/// The context type passed to every call to [`HandlerAction::step`] that provides access to the
+/// underlying. Some of the methods on this type are not intended for use in use supplied handler
```
Comment (SirCipher) [inline server/swimos_agent/src/event_handler/mod.rs:108]:
```suggestion
/// underlying. Some of the methods on this type are not intended for use in supplied handler
```

File: server/swimos_agent/src/event_handler/mod.rs [hunk 11]
```diff
@@ -306,30 +340,41 @@ where
...
 
-/// Error type for fallible event handlers.
+/// Error type for fallible [`HandlerAction`]s. A handler produces an error when a fatal problem occurs and it
+/// cannot produce its result. In most cases this will result in the agent terminating.
 #[derive(Debug, Error)]
 pub enum EventHandlerError {
+    /// Handlers can only be used once. If a handler is stepped after it produces its value, this error will be raised.
     #[error("Event handler stepped after completion.")]
     SteppedAfterComplete,
+    /// An incoming command messages was invalid for the lane it was targetting.
```
Comment (SirCipher) [inline server/swimos_agent/src/event_handler/mod.rs:350]:
```suggestion
    /// An incoming command message was invalid for the lane it was targetting.
```

File: server/swimos_agent/src/event_handler/mod.rs [hunk 12]
```diff
@@ -306,30 +340,41 @@ where
...
     #[error("Invalid incoming message: {0}")]
     BadCommand(AsyncParseError),
+    /// An incoming command messages was incomplete and could not be deserialized.
```
Comment (SirCipher) [inline server/swimos_agent/src/event_handler/mod.rs:353]:
```suggestion
    /// An incoming command message was incomplete and could not be deserialized.
```

File: server/swimos_agent/src/event_handler/mod.rs [hunk 13]
```diff
@@ -306,30 +340,41 @@ where
...
     #[error("An incoming message was incomplete.")]
     IncompleteCommand,
+    /// An error occurred in the agent runtime which prevented this handler from producing its result.
     #[error("An error occurred in the agent runtime.")]
     RuntimeError(#[from] AgentRuntimeError),
+    /// A handler requested a join lane lifecycle with different type parameters than were used to register it.
     #[error("Invalid key or value type for a join lane lifecycle.")]
     BadJoinLifecycle(DowncastError),
+    /// The `on_cue`` lifecycle handler is mandatory for demand lanes. If it is not defined this error will be raised.
```
Comment (SirCipher) [inline server/swimos_agent/src/event_handler/mod.rs:362]:
```suggestion
    /// The `on_cue` lifecycle handler is mandatory for demand lanes. If it is not defined this error will be raised.
```

File: server/swimos_agent/src/event_handler/suspend/mod.rs [hunk 14]
```diff
@@ -114,6 +125,17 @@ where
...
+/// [`EventHandler`] returned by the provided iterator, the handler is scheduled to run after the delay.
+/// The handlers are scheduled sequentially, not simultaneously.
+///
+/// # Note
+///
+/// Both the iterator and the handlers must be [`Send`] as the task running the agent could be moved while
+/// they are still in use.
+///
+/// # Arguments
+/// * `sequence` - An iterator returning a sequence of pairs of delays and handlers.
```
Comment (SirCipher) [inline server/swimos_agent/src/event_handler/suspend/mod.rs:138]:
```suggestion
/// * `schedule` - An iterator returning a sequence of pairs of delays and handlers.
```

File: server/swimos_agent/src/event_handler/suspend/mod.rs [hunk 15]
```diff
@@ -133,6 +155,17 @@ where
...
+/// [`EventHandler`] returned by the provided async [`Stream`], the handler is scheduled to run after the delay.
+/// The handlers are scheduled sequentially, not simultaneously.
+///
+/// # Note
+///
+/// Both the stream and the handlers must be [`Send`] as the task running the agent could be moved while
+/// they are still in use.
+///
+/// # Arguments
+/// * `sequence` - A asynchronous stream returning a sequence of pairs of delays and handlers.
```
Comment (SirCipher) [inline server/swimos_agent/src/event_handler/suspend/mod.rs:168]:
```suggestion
/// * `schedule` - A asynchronous stream returning a sequence of pairs of delays and handlers.
```

File: server/swimos_agent/src/event_handler/suspend/mod.rs [hunk 16]
```diff
@@ -133,6 +155,17 @@ where
     }
 }
 
+/// Schedule a sequence of [`EventHandler`]s to run on a schedule. For each pair of a delay and and
```
Comment (SirCipher) [inline server/swimos_agent/src/event_handler/suspend/mod.rs:158]: The stream doesn't produce a tuple, it's just a handler?

File: server/swimos_agent/src/lib.rs [hunk 17]
```diff
@@ -12,30 +12,76 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//! # Agent Model and Lifecycle Definition API
+//!
+//! This crate provides a framework for building agent implementations that can be executed by the
```
Comment (SirCipher) [inline server/swimos_agent/src/lib.rs:17]:
```suggestion
//! This crate provides a framework for building agent implementations that are executed by the
```

File: server/swimos_agent/src/lib.rs [hunk 18]
```diff
@@ -12,30 +12,76 @@
...
+//! An agent lifecycle is tied to a specific type of agent (any implementation of [`agent_model::AgentSpec`]).
+//! It adds behaviour to the agent by attaching [event handlers](`event_handler::EventHandler`) to various lifecycle
+//! events defined for the agent and its lanes and stores. As the agent is executed by the runtime, each time one of these
+//! events is encountered, the appropriate event handler will be selected by the lifecycle and then executed.
+//!
+//! Multiple lifecycle implementations maybe defined for the same agent specification, allowing the same structure
+//! to be used with different behaviour on separate agent routes.
+
+/// Defines the [agent lifecycle](`agent_lifecycle::AgentLifecycle`) trait, use to specify lifecycles that can be associated with an
+/// agent specification..
```
Comment (SirCipher) [inline server/swimos_agent/src/lib.rs:34]:
```suggestion
/// agent specification.
```

File: server/swimos_agent/src/lib.rs [hunk 19]
```diff
@@ -12,30 +12,76 @@
...
+
+/// Defines the [`event_handler::HandlerAction`] and [`event_handler::EventHandler`] traits used to specify event handlers in agent lifecycle.
 pub mod event_handler;
+
 mod event_queue;
-pub mod item;
+mod item;
+
+/// Defines the lane types that can be included in agent specifications. Lanes are exposed externally by the runtime,
+/// using the Warp protocol (or HTTP in the case of [HTTP lanes](`lanes::HttpLane`)). The states of lanes may be stored
```
Comment (SirCipher) [inline server/swimos_agent/src/lib.rs:54]:
```suggestion
/// using the WARP protocol (or HTTP in the case of [HTTP lanes](`lanes::HttpLane`)). The states of lanes may be stored
```

File: server/swimos_agent/src/stores/map/mod.rs [hunk 20]
```diff
@@ -362,7 +362,7 @@ where
     }
 }
 
-/// An [`EventHandler`] that will alter an entry in the map.
+///  An [event handler](crate::event_handler::EventHandler)`] that will alter an entry in the map.
```
Comment (SirCipher) [inline server/swimos_agent/src/stores/map/mod.rs:365]:
```suggestion
///  An [event handler](crate::event_handler::EventHandler)`] that may alter an entry in the map.
```
As the handler might not do anything

File: server/swimos_introspection/src/forest/mod.rs [hunk 21]
```diff
@@ -137,37 +142,6 @@ impl<D> UriForest<D> {
...
-                        }
-                        current_node.data.as_ref()
-                    }
-                    None => None,
-                }
-            }
-            None => None,
-        }
-    }
-
```
Comment (SirCipher) [inline server/swimos_introspection/src/forest/mod.rs:170]: Why was this removed?

File: server/swimos_introspection/src/forest/tests.rs [hunk 22]
```diff
@@ -82,56 +53,6 @@ fn iters() {
...
-    assert_eq!(
-        uris,
-        HashSet::from([
-            "/unit/1/a/d/e".to_string(),
-            "/unit/1/a/b/c".to_string(),
-            "/unit/1/blah".to_string(),
-        ])
-    );
-}
-
```
Comment (SirCipher) [inline server/swimos_introspection/src/forest/tests.rs:134]: Why are these tests removed?

File: server/swimos_introspection/src/task/mod.rs [hunk 23]
```diff
@@ -111,6 +116,46 @@ pub fn init_introspection(
     (resolver, meta_agent, task)
 }
 
+/// Trait for routing tables with which the introspection agents can be registered.
+pub trait AgentRegistration {
+    /// Register an agent type.
+    ///
+    /// # Arguments
+    /// * `pattern` - The route pattern at which the agent will be loaded.
+    /// * `agent` - The agent type.
```
Comment (SirCipher) [inline server/swimos_introspection/src/task/mod.rs:125]:
```suggestion
    /// * `agent` - The agent to register.
```
As it's accepting an instance.

**Review by SirCipher — CHANGES_REQUESTED** [review]

File: server/swimos_agent/src/lib.rs (hunk 19 above)
Comment (horned-sphere) [reply server/swimos_agent/src/lib.rs:54]: Does it stand for something?

File: server/swimos_introspection/src/forest/mod.rs (hunk 21 above)
Comment (horned-sphere) [reply server/swimos_introspection/src/forest/mod.rs:170]: I moved into in here form the utilities crate and made it private. It didn't seem like it was likely to be used elsewhere. As the visibility got restricted, there were big chunks of it that weren't being used by the introspection so got removed.

File: server/swimos_agent/src/lib.rs (hunk 19 above)
Comment (SirCipher) [reply server/swimos_agent/src/lib.rs:54]: @horned-sphere Web Agent Remote Protocol

File: server/swimos_introspection/src/forest/tests.rs (hunk 22 above)
Comment (horned-sphere) [reply server/swimos_introspection/src/forest/tests.rs:134]: The `prefix_iter` method got removed so its tests went too.

**Review by SirCipher — APPROVED** [review]
